// @lwc/compiler/dist/index.cjs.js
/**
 * Copyright (c) 2026 Salesforce, Inc.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');
var shared = require('@lwc/shared');
var errors = require('@lwc/errors');
var ssrCompiler = require('@lwc/ssr-compiler');
var styleCompiler = require('@lwc/style-compiler');
var templateCompiler = require('@lwc/template-compiler');
var babel = require('@babel/core');
var babelAsyncGeneratorFunctionsPlugin = require('@babel/plugin-transform-async-generator-functions');
var babelAsyncToGenPlugin = require('@babel/plugin-transform-async-to-generator');
var babelClassPropertiesPlugin = require('@babel/plugin-transform-class-properties');
var babelObjectRestSpreadPlugin = require('@babel/plugin-transform-object-rest-spread');
var lockerBabelPluginTransformUnforgeables = require('@locker/babel-plugin-transform-unforgeables');
var lwcClassTransformPlugin = require('@lwc/babel-plugin-component');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var path__namespace = /*#__PURE__*/_interopNamespaceDefault(path);
var styleCompiler__namespace = /*#__PURE__*/_interopNamespaceDefault(styleCompiler);
var babel__namespace = /*#__PURE__*/_interopNamespaceDefault(babel);

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Flag indicating that a warning about still using the deprecated `enableLwcSpread`
 * compiler option has already been logged to the `console`.
 */
let alreadyWarnedAboutLwcSpread = false;
/**
 * Flag indicating that a warning about still using the deprecated `stylesheetConfig`
 * compiler option has already been logged to the `console`.
 */
let alreadyWarnedOnStylesheetConfig = false;
const DEFAULT_OPTIONS = {
    isExplicitImport: false,
    preserveHtmlComments: false,
    enableStaticContentOptimization: true,
    // TODO [#3370]: remove experimental template expression flag
    experimentalComplexExpressions: false,
    disableSyntheticShadowSupport: false,
    enableLightningWebSecurityTransforms: false,
    targetSSR: false,
    ssrMode: shared.DEFAULT_SSR_MODE,
    experimentalErrorRecoveryMode: false,
    componentFeatureFlagModulePath: '',
};
const DEFAULT_DYNAMIC_IMPORT_CONFIG = {
    loader: '',
    strictSpecifier: true,
};
const DEFAULT_STYLESHEET_CONFIG = {
    customProperties: {
        resolution: { type: 'native' },
    },
};
const DEFAULT_OUTPUT_CONFIG = {
    minify: false,
    sourcemap: false,
};
/**
 * Validates that the options conform to the expected shape and normalizes them to a standard format
 * @param options Input options
 * @returns Normalized options
 * @example
 * const normalizedOptions = validateTransformOptions({
 *   namespace: 'c',
 *   name: 'app',
 * })
 */
function validateTransformOptions(options) {
    validateOptions(options);
    return normalizeOptions(options);
}
function validateOptions(options) {
    errors.invariant(!shared.isUndefined(options), errors.CompilerValidationErrors.MISSING_OPTIONS_OBJECT, [options]);
    if (!shared.isUndefined(options.enableLwcSpread) && !alreadyWarnedAboutLwcSpread) {
        alreadyWarnedAboutLwcSpread = true;
        // eslint-disable-next-line no-console
        console.warn(`"enableLwcSpread" property is deprecated. The value doesn't impact the compilation and can safely be removed.`);
    }
    if (!shared.isUndefined(options.stylesheetConfig) && !alreadyWarnedOnStylesheetConfig) {
        alreadyWarnedOnStylesheetConfig = true;
        // eslint-disable-next-line no-console
        console.warn(`"stylesheetConfig" property is deprecated. The value doesn't impact the compilation and can safely be removed.`);
    }
    if (!shared.isUndefined(options.outputConfig)) {
        validateOutputConfig(options.outputConfig);
    }
}
function isUndefinedOrBoolean(property) {
    return shared.isUndefined(property) || shared.isBoolean(property);
}
function validateOutputConfig(config) {
    errors.invariant(isUndefinedOrBoolean(config.sourcemap) || config.sourcemap === 'inline', errors.CompilerValidationErrors.INVALID_SOURCEMAP_PROPERTY, [config.sourcemap]);
    if (!shared.isUndefined(config.minify)) {
        // eslint-disable-next-line no-console
        console.warn(`"OutputConfig.minify" property is deprecated. The value doesn't impact the compilation and can safely be removed.`);
    }
}
function normalizeOptions(options) {
    const outputConfig = {
        ...DEFAULT_OUTPUT_CONFIG,
        ...options.outputConfig,
    };
    const stylesheetConfig = {
        customProperties: {
            ...DEFAULT_STYLESHEET_CONFIG.customProperties,
            ...(options.stylesheetConfig && options.stylesheetConfig.customProperties),
        },
    };
    const experimentalDynamicComponent = {
        ...DEFAULT_DYNAMIC_IMPORT_CONFIG,
        ...options.experimentalDynamicComponent,
    };
    const apiVersion = shared.getAPIVersionFromNumber(options.apiVersion);
    return {
        ...DEFAULT_OPTIONS,
        ...options,
        stylesheetConfig,
        outputConfig,
        experimentalDynamicComponent,
        apiVersion,
    };
}

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transform the passed source code
 * @param src The source to be transformed. Can be the content of a JavaScript, HTML, or CSS file.
 * @param filename The source filename, with extension.
 * @param config The transformation options. The `name` and the `namespace` of the component is the
 * minimum required for transformation.
 * @returns An object with the generated code, source map and gathered metadata.
 * @throws Compilation errors
 */
function styleTransform(src, filename, config) {
    const { customProperties } = config.stylesheetConfig;
    const { experimentalErrorRecoveryMode } = config;
    const styleCompilerConfig = {
        customProperties: {
            resolverModule: customProperties.resolution.type === 'module'
                ? customProperties.resolution.name
                : undefined,
        },
        scoped: config.scopedStyles,
        disableSyntheticShadowSupport: config.disableSyntheticShadowSupport,
        apiVersion: config.apiVersion,
        experimentalErrorRecoveryMode,
    };
    let res;
    try {
        res = styleCompiler__namespace.transform(src, filename, styleCompilerConfig);
    }
    catch (e) {
        // Handle AggregateError when in error recovery mode
        if (experimentalErrorRecoveryMode && e instanceof AggregateError) {
            const compilerErrors = e.errors.map((error) => errors.normalizeToCompilerError(errors.TransformerErrors.CSS_TRANSFORMER_ERROR, error, {
                filename,
            }));
            throw new errors.CompilerAggregateError(compilerErrors, 'Multiple CSS errors occurred during compilation.');
        }
        throw errors.normalizeToCompilerError(errors.TransformerErrors.CSS_TRANSFORMER_ERROR, e, { filename });
    }
    // Rollup only cares about the mappings property on the map. Since producing a source map for
    // the styles doesn't make sense, the transform returns an empty mappings.
    return {
        code: res.code,
        map: { mappings: '' },
    };
}

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transforms a HTML template into module exporting a template function.
 * The transform also add a style import for the default stylesheet associated with
 * the template regardless if there is an actual style or not.
 * @param src HTML source
 * @param filename Source filename, with extension.
 * @param options Transformation options
 * @returns Transformed code, source map, and metadata
 * @throws Compiler errors, when compilation fails.
 * @example
 */
function templateTransform(src, filename, options) {
    const { experimentalDynamicComponent, 
    // TODO [#3370]: remove experimental template expression flag
    experimentalComplexExpressions, preserveHtmlComments, enableStaticContentOptimization, customRendererConfig, enableDynamicComponents, experimentalDynamicDirective: deprecatedDynamicDirective, enableLwcOn, instrumentation, namespace, name, apiVersion, disableSyntheticShadowSupport, experimentalErrorRecoveryMode, } = options;
    const experimentalDynamicDirective = deprecatedDynamicDirective ?? Boolean(experimentalDynamicComponent);
    let result;
    try {
        result = templateCompiler.compile(src, filename, {
            name,
            namespace,
            experimentalDynamicDirective,
            // TODO [#3370]: remove experimental template expression flag
            experimentalComplexExpressions,
            preserveHtmlComments,
            enableStaticContentOptimization,
            customRendererConfig,
            enableDynamicComponents,
            enableLwcOn,
            instrumentation,
            apiVersion,
            disableSyntheticShadowSupport,
        });
    }
    catch (e) {
        throw errors.normalizeToCompilerError(errors.TransformerErrors.HTML_TRANSFORMER_ERROR, e, { filename });
    }
    const errors$1 = result.warnings.filter((warning) => warning.level === errors.DiagnosticLevel.Error);
    if (experimentalErrorRecoveryMode && errors$1.length > 0) {
        throw new errors.CompilerAggregateError(errors$1.map((err) => errors.CompilerError.from(err, { filename })));
    }
    if (errors$1[0]) {
        throw errors.CompilerError.from(errors$1[0], { filename });
    }
    // The "Error" diagnostic level makes no sense to include here, because it would already have been
    // thrown above. As for "Log" and "Fatal", they are currently unused.
    const warnings = result.warnings.filter((_) => _.level === errors.DiagnosticLevel.Warning);
    // Rollup only cares about the mappings property on the map. Since producing a source map for
    // the template doesn't make sense, the transform returns an empty mappings.
    return {
        code: result.code,
        map: { mappings: '' },
        warnings,
        cssScopeTokens: result.cssScopeTokens,
    };
}

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transforms a JavaScript file.
 * @param code The source code to transform
 * @param filename The source filename, with extension.
 * @param options Transformation options.
 * @returns Compiled code
 * @throws Compilation errors
 * @example
 */
function scriptTransform(code, filename, options) {
    const { isExplicitImport, enableSyntheticElementInternals, 
    // TODO [#5031]: Unify dynamicImports and experimentalDynamicComponent options
    experimentalDynamicComponent: dynamicImports, outputConfig: { sourcemap }, enableLightningWebSecurityTransforms, namespace, name, instrumentation, apiVersion, experimentalErrorRecoveryMode, componentFeatureFlagModulePath, } = options;
    const lwcBabelPluginOptions = {
        isExplicitImport,
        dynamicImports,
        enableSyntheticElementInternals,
        namespace,
        name,
        instrumentation,
        apiVersion,
        componentFeatureFlagModulePath,
    };
    const plugins = [
        [lwcClassTransformPlugin, lwcBabelPluginOptions],
        [babelClassPropertiesPlugin, { loose: true }],
    ];
    if (!shared.isAPIFeatureEnabled(3 /* APIFeature.DISABLE_OBJECT_REST_SPREAD_TRANSFORMATION */, apiVersion)) {
        plugins.push(babelObjectRestSpreadPlugin);
    }
    if (enableLightningWebSecurityTransforms) {
        plugins.push(lockerBabelPluginTransformUnforgeables, babelAsyncToGenPlugin, babelAsyncGeneratorFunctionsPlugin);
    }
    let result;
    try {
        result = babel__namespace.transformSync(code, {
            filename,
            sourceMaps: sourcemap,
            // Prevent Babel from loading local configuration.
            babelrc: false,
            configFile: false,
            // Force Babel to generate new line and white spaces. This prevent Babel from generating
            // an error when the generated code is over 500KB.
            compact: false,
            plugins,
            parserOpts: {
                errorRecovery: experimentalErrorRecoveryMode,
            },
        });
    }
    catch (e) {
        // If we are here in errorRecoveryMode then it's most likely that we have run into
        // an unforeseen error
        let transformerError = errors.TransformerErrors.JS_TRANSFORMER_ERROR;
        // Sniff for a Babel decorator error, so we can provide a more helpful error message.
        if (e.code === 'BABEL_TRANSFORM_ERROR' &&
            e.message?.includes('Decorators are not enabled.') &&
            /\b(track|api|wire)\b/.test(e.message) // sniff for @track/@api/@wire
        ) {
            transformerError = errors.TransformerErrors.JS_TRANSFORMER_DECORATOR_ERROR;
        }
        throw errors.normalizeToCompilerError(transformerError, e, { filename });
    }
    if (experimentalErrorRecoveryMode) {
        const metadata = result.metadata;
        const errors$1 = metadata?.lwcErrors;
        if (errors$1) {
            throw new errors.CompilerAggregateError(errors$1.map((diagnostic) => errors.CompilerError.from(diagnostic)), 'Multiple errors occurred during compilation.');
        }
    }
    return {
        code: result.code,
        map: result.map,
    };
}

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transform the passed source code.
 * @param src The source to be transformed. Can be the content of a JavaScript, HTML, or CSS file.
 * @param filename The source filename, with extension.
 * @param options The transformation options. The `name` and the `namespace` of the component is the
 * minimum required for transformation.
 * @returns A promise resolving to an object with the generated code, source map and gathered metadata.
 * @example
 * const source = `
 *     import { LightningElement } from 'lwc';
 *     export default class App extends LightningElement {}
 * `;
 * const filename = 'app.js';
 * const options = {
 *     namespace: 'c',
 *     name: 'app',
 * };
 * const { code } = await transform(source, filename, options);
 * @deprecated Use {@linkcode transformSync} instead
 */
function transform(src, filename, options) {
    validateArguments(src, filename);
    return new Promise((resolve, reject) => {
        try {
            const res = transformSync(src, filename, options);
            resolve(res);
        }
        catch (error) {
            reject(error);
        }
    });
}
/**
 * Transform the passed source code
 * @param src The source to be transformed. Can be the content of a JavaScript, HTML, or CSS file.
 * @param filename The source filename, with extension.
 * @param options The transformation options. The `name` and the `namespace` of the component is the
 * minimum required for transformation.
 * @returns An object with the generated code, source map and gathered metadata.
 * @example
 *
 * const source = `
 *     import { LightningElement } from 'lwc';
 *     export default class App extends LightningElement {}
 * `;
 * const filename = 'app.js';
 * const options = {
 *     namespace: 'c',
 *     name: 'app',
 * };
 * const { code } = transformSync(source, filename, options);
 */
function transformSync(src, filename, options) {
    validateArguments(src, filename);
    const normalizedOptions = validateTransformOptions(options);
    return transformFile(src, filename, normalizedOptions);
}
function validateArguments(src, filename) {
    errors.invariant(shared.isString(src), errors.TransformerErrors.INVALID_SOURCE, [src]);
    errors.invariant(shared.isString(filename), errors.TransformerErrors.INVALID_ID, [filename]);
}
function transformFile(src, filename, options) {
    switch (path__namespace.extname(filename)) {
        case '.html':
            if (options.targetSSR) {
                return ssrCompiler.compileTemplateForSSR(src, filename, options, options.ssrMode);
            }
            return templateTransform(src, filename, options);
        case '.css':
            return styleTransform(src, filename, options);
        case '.tsx':
        case '.jsx':
        case '.ts':
        case '.js':
        case '.mts':
        case '.mjs':
            if (options.targetSSR) {
                return ssrCompiler.compileComponentForSSR(src, filename, options, options.ssrMode);
            }
            return scriptTransform(src, filename, options);
        default:
            throw errors.generateCompilerError(errors.TransformerErrors.NO_AVAILABLE_TRANSFORMER, {
                messageArgs: [filename],
                origin: { filename },
            });
    }
}

/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/** The version of LWC being used. */
const version = "8.28.2";

exports.transform = transform;
exports.transformSync = transformSync;
exports.version = version;
/** version: 8.28.2 */
